### ISSUE TITLE
When doing master failover, there is a race between starting WConfd and updating the configuration

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1171]
Originally reported of Google Code with ID 1115.

```
<b>What steps will reproduce the problem?</b>
1. Run master failover on a master candidate

<b>What is the expected output? What do you see instead?</b>

Occasionally it happens that the temporary WConfd instance on the new master doesn't start fast enough and the request for updating the configuration fails.

And since we're not running on the master node, the Transport class doesn't retry. The proper fix is to modify this particular call so that the Transport class properly retries, even if not on the master node.


```
Originally added on 2015-07-01 16:47:22 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
See the discussion at https://groups.google.com/d/topic/ganeti/lTSmTqRPmu0/discussion

```
Originally added on 2015-07-02 13:43:20 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The same problem with the Transport class was reported for Metad as well: https://groups.google.com/d/topic/ganeti/A1zcEZNliAo/discussion
```
Originally added on 2015-07-03 08:09:04 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed in (for WConfd as well as Metad):

commit 7e16570b6de06c4ade8735c03b0d4af786c4bd46
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 15:27:13 2015 +0200

    When connecting to Metad fails, log the full stack trace
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit f2d96bc3defdd541f8ad9fcd5508c886b5c7c7f1
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 15:11:00 2015 +0200

    Set up the Metad client with allow_non_master
    
    .. since the communication takes place on non-master nodes.
    
    This ensures the client properly retries if there is a communication
    failure.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit aa27b56a086d2da33f3e9eb1c9893c4f2d0a7da9
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 11:29:28 2015 +0200

    Set up the configuration client properly on non-masters
    
    If the configuration client is opened in the 'accept_foreign' mode,
    meaning it is running on a non-master node temporarily, the option
    needs to be propagated to the RPC client as well.
    
    This fixes issue #1115.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit f828b17a7166e1261dd44da1d255811d94806201
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 11:28:41 2015 +0200

    Add the 'allow_non_master' option to the WConfd RPC client
    
    While at it, fix the call to the AbstractStubClient to properly pass the
    keyword arguments.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 706952c1759b2776d7c64ddaaa496b2860cfb5c4
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 11:26:18 2015 +0200

    Add the option to disable master checks to the RPC client
    
    The option is propagated to the Transport class and allows to disable
    checks for the master node, if the client is run on a different node on
    purpose.
    
    While at it, fix the documentation for the arguments of the constructors
    of the classes.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 11efe0ff5c8b28dffe02fbeb84b8587e90a058d8
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 13:51:34 2015 +0200

    Add 'allow_non_master' to the Luxi test transport class too
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 17c0f0d47a12420c13f20b85825f7b73f7fa77c2
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 11:27:13 2015 +0200

    Add 'allow_non_master' to FdTransport for compatibility
    
    Since it serves as an alternative to the Transport class, it should
    support the same constructor options, even if it doesn't use them.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit 1d464c1ce1b6d746473a4bbd42c4e94af08a0d62
Author: Petr Pudlak <pudlak@google.com>
Date:   Thu Jul 2 11:58:02 2015 +0200

    Properly document all constructor arguments of Transport
    
    .. while documenting allow_non_master.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

commit fa255d206e2c8a0acaf52f12c238b396ee1fe039
Author: Petr Pudlak <pudlak@google.com>
Date:   Fri Jun 5 14:13:48 2015 +0200

    Allow the Transport class to be used for non-master nodes
    
    If a communication failure occurred and the caller was not running on
    the master node, Transport assumed that this itself was the cause of
    the error condition.
    
    However, for communication with the metadata daemon we need to support
    non-master nodes as well.
    
    Add a parameter that allows to use the class on non-master nodes.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>
    
    Cherry-picked-from: ade70feb258a57ae0565395ba48ac2b3ef02b1c0
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Helga Velroyen <helgav@google.com>

```
Originally added on 2015-07-03 10:56:15 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2015-07-13 17:13:17 +0000 UTC.

Changed State: Released
